File: products-service/src/main/java/com/arielzarate/products/interfaces/errors/exception/ExternalClientException.java
package com.arielzarate.products.interfaces.errors.exception;

import com.arielzarate.products.interfaces.errors.model.ClientError;
import lombok.Getter;

@Getter
public class ExternalClientException extends RuntimeException {

    private final ClientError clientError;

    public ExternalClientException(ClientError clientError) {
        super("Error calling external client: " + clientError.title()); //message
        this.clientError = clientError;
    }

}

File: shipment/src/main/kotlin/com/arielzarate/shipment/interfaces/error/exception/ExternalClientException.kt
package com.arielzarate.shipment.interfaces.error.exception

import com.arielzarate.shipment.interfaces.error.model.ClientError
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR

class ExternalClientException(val error: ClientError) : RuntimeException(){
    companion object {
        fun timeout(uri: String, description: String): ClientError {
            return ClientError(
                title = "Timeout calling $uri",
                detail = description,
                status = INTERNAL_SERVER_ERROR.value(),
                type = null,
                instance = null,
                errors = null
            )
        }

        fun error(uri: String, description: String): ClientError {
            return ClientError(
                title = "Error with $uri",
                detail = description,
                status = INTERNAL_SERVER_ERROR.value(),
                type = null,
                instance = null,
                errors = null
            )
        }
    }
}
